contracts/Core/StakeManager.sol

```diff
@@ -172,44 +175,65 @@ contract StakeManager is Initializable, StakeStorage, StateManager, Pause, Stake
     /// @param stakerId The Id of staker associated with sRZR which user want to unstake
     /// @param sAmount The Amount in sRZR
     function unstake(uint32 stakerId, uint256 sAmount) external initialized whenNotPaused {
+        require(sAmount > 0, "Non-Positive Amount");
+        require(stakerId != 0, "staker.id = 0");
+        require(stakers[stakerId].stake > 0, "Nonpositive stake");
+        require(locks[msg.sender][stakers[stakerId].tokenAddress][LockType.Unstake].amount == 0, "Existing Unstake Lock");
+
+        uint32 epoch = _getEpoch(epochLength);
+        Structs.Staker storage staker = stakers[stakerId];
+        IStakedToken sToken = IStakedToken(staker.tokenAddress);
+
+        require(sToken.balanceOf(msg.sender) >= sAmount, "Invalid Amount");
+
+        locks[msg.sender][staker.tokenAddress][LockType.Unstake] = Structs.Lock(
+            sAmount,
+            0,
+            epoch + unstakeLockPeriod,
+            sToken.getRZRDeposited(msg.sender, sAmount)
+        );
+        emit Unstaked(msg.sender, epoch, stakerId, sAmount, staker.stake, block.timestamp);
+        require(sToken.transferFrom(msg.sender, address(this), sAmount), "sToken transfer failed");
+    }
+
+    function initiateWithdraw(uint32 stakerId) external initialized whenNotPaused {
         State currentState = _getState(epochLength);
         require(currentState != State.Propose, "Unstake: NA Propose");
         require(currentState != State.Dispute, "Unstake: NA Dispute");
-        uint32 epoch = _getEpoch(epochLength);
 
+        require(stakerId != 0, "staker doesnt exist");
+        uint32 epoch = _getEpoch(epochLength);
         Structs.Staker storage staker = stakers[stakerId];
-        require(staker.id != 0, "staker.id = 0");
-        require(staker.stake > 0, "Nonpositive stake");
-        require(locks[msg.sender][staker.tokenAddress].amount == 0, "Existing Lock");
-        require(sAmount > 0, "Non-Positive Amount");
+        Structs.Lock storage lock = locks[msg.sender][staker.tokenAddress][LockType.Unstake];
+        require(lock.unlockAfter != 0, "Did not unstake");
+        require(lock.unlockAfter <= epoch, "Withdraw epoch not reached");
+        require(lock.unlockAfter + withdrawInitiationPeriod >= epoch, "Initiation Period Passed"); // Can Use ExtendLock
```

Comment:
hey as of there looks like three lock periods here, unstakeLockPeriod, withdrawInitiationPeriod, and withdrawLockPeriod. If I understand correctly there is unstake, then unlockAfter is = epoch + unstakeLockPeriod. then initiaite withdraw can only happen if epoch<= epoch + unstakeLockPeriod + withdrawInititationPeriod, right ? If yes please consider removing any one of it as its redundant

Reply:
how is it redundant? withdrawInitiationPeriod is similar to what we had with withdrawReleasePeriod we have currently

Reply:
hmm maybe I am missing something, lets check this in todays call

Reply:
Is it like this 
unstakeLockPeriod, : 1
withdrawInitiationPeriod, : 5
withdrawLockPeriod. : 1


epoch 0 
unstake 
unLockAfter : 1

only allow initiate withdraw when unlockAfter + 5 <= epoch 

now my question is in unstake only, couldnt we have assigned this unlockAfter + 5 value only
6 rather than 1

Reply:
lets discuss this anyway on call today
as I am not sure if this is needed now